preview_generator/preview/builder/document_generic.py

```diff
@@ -0,0 +1,209 @@
+        raise NotImplementedError
+
+    def _cache_file_process_already_running(self, file_name: str) -> bool:
+        if os.path.exists(file_name + '_flag'):
+            return True
+        else:
+            return False
+
+    def build_jpeg_preview(
+        self,
+        file_path: str,
+        preview_name: str,
+        cache_path: str,
+        page_id: int,
+        extension: str='.jpg',
+        size: ImgDims=None,
+        attempt=0
+    ) -> None:
+
+        cache_file = os.path.join(cache_path, preview_name)
+
+        if self._cache_file_process_already_running(cache_file):
+            # Note - Basile - infinite recursion protection
+            if attempt >= 5:
+                raise PreviewGeneratorException(
+                    'Max attempts exceeded aborting preview'
+                )
+            attempt += 1
+            time.sleep(2)
+            return self.build_jpeg_preview(
+                file_path=file_path, preview_name=preview_name,
+                cache_path=cache_path, extension=extension, page_id=page_id
+            )
+
+        input_pdf_stream = None
+        if os.path.exists(os.path.join(cache_path, preview_name + '.pdf')):
+            input_pdf_stream = open(
+                os.path.join(cache_path, preview_name + '.pdf'), 'rb'
+            )
+
+        if not input_pdf_stream:
+            with open(file_path, 'rb') as _file:
+                file, file_extension = os.path.splitext(file_path)
+                input_pdf_stream = self._convert_to_pdf(
+                    _file, file_extension, cache_path, preview_name
+                )
+
+        input_pdf = PdfFileReader(input_pdf_stream)
+        intermediate_pdf = PdfFileWriter()
+        intermediate_pdf.addPage(input_pdf.getPage(int(page_id)))
+
+        intermediate_pdf_stream = BytesIO()
+        intermediate_pdf.write(intermediate_pdf_stream)
+        intermediate_pdf_stream.seek(0, 0)
+        jpeg_stream = convert_pdf_to_jpeg(intermediate_pdf_stream, size)
+
+        jpeg_preview_path = os.path.join(cache_path, preview_name + extension)
+        with open(jpeg_preview_path, 'wb') as jpeg_output_stream:
+            buffer = jpeg_stream.read(1024)
+            while buffer:
+                jpeg_output_stream.write(buffer)
+                buffer = jpeg_stream.read(1024)
+
+    def build_pdf_preview(
+        self,
+        file_path: str,
+        preview_name: str,
+        cache_path: str,
+        extension: str = '.pdf',
+        page_id: int = -1
+    ) -> None:
+
+        input_extension = os.path.splitext(file_path)[1]
+        intermediate_pdf_filename = preview_name.split('-page')[0] + '.pdf'
+        intermediate_pdf_file_path = os.path.join(
+            cache_path,
+            intermediate_pdf_filename
+        )
+
+        if not os.path.exists(intermediate_pdf_file_path):
+            if os.path.exists(intermediate_pdf_file_path + '_flag'):
+                # Wait 2 seconds, then retry
+                time.sleep(2)
```

Comment:
Why waiting for 2 sec ? This may cause a process who use preview_generator to hang more than 2 second ?
An "INFO" comment may be nice here explaining the goal of this.

Reply:
you're right that an explanation is a good idea.
The use case is the following one: 

you ask for a jpeg preview of a libreoffice document. the first step is to convert odt to full pdf. If at the same time you ask for a jpeg preview of another page, then the full pdf will be geenrated twice, resulting in a corrupt file (and loss of time). Waiting 2 seconds let the running transformation to finish (the 2s delay is only executed in case a preview is already running)

@basilelegal could you add a comment explaining this?

Reply:
Author of this code is not @basilelegal (it's a move of code). But, maybe he is able to respond ?

Reply:
It's coming with the next push

Reply:
up

Reply:
done